### PR TITLE
Let BnsConnection.postTx resolve before block

### LINKED ISSUE #564: Let postTx in BNS resolve before block
As discussed in https://github.com/iov-one/iov-core/pull/561#discussion_r236291870, it is more consistent to resolve postTx after sucessfully sending a transaction to mempool.

> watchTx based on hash will get an event when it is writen to a block.
> 
> If you subscribe before the commit, you will get notified when it is inside. Then you can listen to blockheaders to increment the confirmation.
> 
> This is just a change to the blocking call up to the first confirmation

This requires an additional

```ts
await response.blockInfo.waitFor(info => info.state === BcpTransactionState.InBlock);
```

for all code that expects the transaction to be in block (a lot of tests).

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #564

### CODE REVIEW

File: packages/iov-core/src/multichainsigner.spec.ts [hunk 1]
```diff
@@ -87,7 +85,8 @@ describe("MultiChainSigner", () => {
           tokenTicker: cash,
         },
       };
-      await signer.signAndCommit(sendTx, mainWalletId);
+      const postResponse = await signer.signAndCommit(sendTx, mainWalletId);
```
Comment (LucasIsasmendi) [inline packages/iov-core/src/multichainsigner.spec.ts]: from `SignAndCommit` to `SignAndCommitAndWaitInBlockConfirmation` will be an accurate function name description change

File: packages/iov-lisk/src/liskconnection.ts [hunk 2]
```diff
@@ -93,7 +93,7 @@ export class LiskConnection implements BcpConnection {
     return responseBody.data.height;
   }
 
-  public async postTx(bytes: PostableBytes): Promise<BcpTransactionResponse> {
+  public async postTx(bytes: PostableBytes): Promise<PostTxResponse> {
```
Comment (LucasIsasmendi) [inline packages/iov-lisk/src/liskconnection.ts]: postTx -> postTxAndWaitInBlockConfirmation (sames as in bnsconnection postTx)

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 3]
```diff
@@ -284,7 +286,8 @@ describe("BnsConnection", () => {
       const nonce = await getNonce(connection, faucetAddr);
       const signed = await profile.signTransaction(mainWalletId, faucet, sendTx, bnsCodec, nonce);
       const txBytes = bnsCodec.bytesToPost(signed);
-      await connection.postTx(txBytes);
+      const response = await connection.postTx(txBytes);
+      await response.blockInfo.waitFor(info => info.state === BcpTransactionState.InBlock);
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.spec.ts]: Waiting for the block is not the main thing we test here. We only wait and check balance to verify the send was successful. This test is indeed doing too much at once. I prefer to keep the description as it is and reduce the code of the test.

File: packages/iov-bns/src/bnsconnection.ts [hunk 4]
```diff
@@ -157,53 +157,65 @@ export class BnsConnection implements BcpAtomicSwapConnection {
...
-      confirmations: 1,
-    };
+    const firstEvent: BcpBlockInfo = { state: BcpTransactionState.Pending };
     let blocksSubscription: Subscription;
     let lastEventSent: BcpBlockInfo = firstEvent;
     const blockInfoProducer = new DefaultValueProducer<BcpBlockInfo>(firstEvent, {
-      onStarted: () => {
-        blocksSubscription = this.changeBlock().subscribe({
-          next: blockHeight => {
+      onStarted: async () => {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.ts]: We don't. All the code of the `blockInfoProducer` only runs when someone starts to listen for the blockInfo updates. If you don't subscribe, the function does nothing more than "post".

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 5]
```diff
@@ -698,9 +722,11 @@ describe("BnsConnection", () => {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.spec.ts]: No, right now we have no interfaces to access the transaction pool

File: packages/iov-core/src/multichainsigner.spec.ts (hunk 1 above)
Comment (webmaster128) [reply packages/iov-core/src/multichainsigner.spec.ts]: same as above: "signAndCommit" does not wait for a block. Only when you subscribe to blockInfo, you will get updates about the block. This is the reason we need to wait for the block separatly one line below.

File: packages/iov-lisk/src/liskconnection.ts (hunk 2 above)
Comment (webmaster128) [reply packages/iov-lisk/src/liskconnection.ts]: This does not wait for a block. But you can track a posted transaction and observe its confirmations if you want like in this example: https://github.com/iov-one/iov-core/blob/master/packages/iov-lisk/src/liskconnection.spec.ts#L295

File: packages/iov-bns/src/bnsconnection.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.ts]: This is what I love about xstream producers: they only produce data if someone cares (i.e. subscribed to a stream), which avoids overhead when no one is listening.

File: packages/iov-bns/src/bnsconnection.spec.ts [hunk 6]
```diff
@@ -698,9 +722,11 @@ describe("BnsConnection", () => {
 
       const nonce = await getNonce(connection, faucetAddress);
       const signed = await profile.signTransaction(mainWalletId, faucet, sendTx, bnsCodec, nonce);
-      const txBytes = bnsCodec.bytesToPost(signed);
-      const response = await connection.postTx(txBytes);
-      const txHeight = response.metadata.height!;
+      const response = await connection.postTx(bnsCodec.bytesToPost(signed));
+      const blockInfo = await response.blockInfo.waitFor(info => info.state === BcpTransactionState.InBlock);
+      const txHeight = (blockInfo as BcpBlockInfoInBlock).height;
+
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.spec.ts]: I split this into two separate tests now, so we have each search type separatly:

```
   5. searchTx
      ✔ can search for transactions by tags (1128ms)
      ✔ can search for transactions by height (1029ms)
      ✔ can search for transactions by hash (1031ms)
      ! can search for transactions by minHeight/maxHeight
```

File: packages/iov-bns/src/bnsconnection.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.spec.ts]: Postponing this issue as out of scope for this PR. We don't change the test behaviour here. We just need two lines now for the functionality that was done by one line before.

File: packages/iov-bns/src/bnsconnection.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (LucasIsasmendi) [reply packages/iov-bns/src/bnsconnection.ts]: I am talking here about the concept of create a function that only does one thing, or a sequence of steps, related with the name of the function.
Inside of `postTx` there is a complete different code that is not related with the concept of post a transaction and neither is reflected in the name of the function. Such functionality can be wrapped inside another function, for example, `watchTxState(txId)`

[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.ts]: Ah, yeah I see. In this case this is not so easy to archive I guess. Ethan [requested that behaviour](https://github.com/iov-one/iov-core/issues/413#issuecomment-437146536) and from a user's perspective it makes sense to post a transaction and being able to track the state of the method's result object. There is also a problem with having two calls postTx/watchTxState: in order to ensure you don't miss events you need to pre-calculate the Transaction ID, start watching first, then post. This makes the caller code much more complicated but the implementation simpler. I guess the current overall aproach is to optimize for nice caller code at the BCP level, which gives us high level interfaces.

What we could do of course is to pull out the watching code in a private helper function to improve readability. But I guess this discussion is about the `postTx` API, right?

**Pull request merged**